### DISCUSSION TITLE
Livewire V3 - Release

### DISCUSSION THREAD

**Discussion by lrljoe** [rappasoft/laravel-livewire-tables#1305] [General]
# Please Note
With the stable release of Livewire 3.0, v3 of this package is now released, the recommendation is notv publish views, as they may be subject to change at short notice to address any issues raised.

Once Livewire v3 is in use, v3 of this package should be used.

----

### Potential Breaking Changes

1. Multiple Relationships to Same Model/Table
If your table component has multiple relationships to the same table, then please see this PR for more information:
[https://github.com/rappasoft/laravel-livewire-tables/pull/1190](https://github.com/rappasoft/laravel-livewire-tables/pull/1190)
This PR has been adjusted and merged into V3.0

2. View/Blade Changes
There have been numerous changes to the blades/views.  If you have previously published the views, then you will experience a breaking change unless you remove the published views.  We recommend not publishing views unless you have a specific use case.  It is almost always possible to make necessary adjustments using the built-in methods!

----

# Configure() New Methods
- Added support for cursor pagination ($this->setPaginationMethod('cursor'))
- Search
  - Search Placeholder is now customisable ($this->setSearchPlaceholder('Finder'))
  - Removed $this->setSearchLazy()
  - Added $this->setSearchBlur() - updates Search when blurring the search input
  - Added $this->setSearchThrottle(int $milliseconds) - throttles updating the search input

# Pagination
- Cursor & Simple pagination both include a "Total Item Count" stored in $paginationTotalItemCount, which is displayed in the table foot

# Bootstrap Striping
Bootstrap striping is now working as standard

# Icons
The package now uses HeroIcons instead of hard-coded SVGs for icons to improve performance and standardise the look & feel.

# Filter Improvements
- DateFilter & DateTimeFilter have customisable Pills Date Format
- DateFilter & DateTimeFilter support setFilterDefaultValue to set a Default Value
- MultiSelectFilter & MultiSelectDropdownFilter both support setFirstOption() to set a value for "All"

## New Filters
There are two new filters:
### NumberRange filter (No 3rd party dependency)
![image](https://github.com/rappasoft/laravel-livewire-tables/assets/104938042/f9b926ca-263a-4bde-8c86-544d63817c61)

### DateRange filter (using Flatpickr)
If you would like to more smoothly filter your query by a start and end date, you can use the new DateRangeFilter:
![image](https://github.com/rappasoft/laravel-livewire-tables/assets/104938042/7f5398cc-00ab-44a6-9361-9974f68b5099)

## JS/CSS Injection
This package uses the same approach as Livewire 3.0 for injecting the relevant JS and CSS files.  It is recommended during the beta that you utilise the injection method for this package, however you can also choose to manually include the files, or reference them in your app.js should you choose.  See the docs for more details
These files are used for:
- Bulk Actions
- Pagination Counts
- Reordering
- Filters
  - Number Range Filter
  - Date Range Filter

It is recommended to allow the injection, or include the files in your build process, even if you are not using any of this functionality, to avoid Alpine errors.

## Reordering
If you are using the reorder functionality, then you no longer require the 3rd party library, and should remove it, as it is not supported with Livewire v3.

# Additional Documentation
The v3 documentation is now available on the Rappasoft Website.

# Issues
When [raising an issue](https://github.com/rappasoft/laravel-livewire-tables/issues/new?assignees=&labels=bug&projects=&template=bug.yml&title=%5BBug%5D%3A+), please make sure to first check for an existing issue before raising a new issue.

Please try to add all relevant information, particularly theme and versions, as this helps massively when trying to replicate an issue.  Screenshots, both of the issue, and any logs are also massively appreciated.

# Feedback
If you are able to, it'd be appreciated if you could let us know that you're testing the beta on the Discord:
[https://discord.gg/mkmEfq7hUC](https://discord.gg/mkmEfq7hUC), and do feel free to ask if you come across something!

We have done our best to make migrating as simple as possible, however there are a couple of changes that you may need to make!

**Comment by lrljoe** [comment]
You should be able to use the "v3-develop" branch now.

Please jump on the discord to ping myself/Rappasoft if anything is immediately seeming odd!

**Comment by lrljoe** [comment]
Updates to Develop:
- Removal of setSearchLazy (not supported in LW3)
- Fix setSearchDebounce to work correctly
- Add setSearchThrottle
- Add setSearchBlur

- Fix Publishing Views
- Initial Fix - Publishing Translations

# New Feature - Append and Prepend Columns
This provides the ability to prepend/append columns to your array.  

Allowing you to use a trait/method in your component (e.g. if you have a standard set of columns that you always want to put on the end/start of your table)
Often you will want to re-use the same column across multiple tables.  For example a "Created At" and "Used At" column.

To mitigate the pain of maintaining this, two new methods have been introduced.

These methods both function in exactly the same way as your standard columns(), and expect an array of columns.

Any columns defined in prependColumns() will be the first columns in your list of columns.
```
public function prependColumns(): array
{
    return [];
}
```

Any columns defined in appendColumns() will be the last columns in your list of columns.
```
public function appendColumns(): array
{
    return [];
}
```

You can call these in your trait, and they will be automatically appended/prepended to tables.

For example, to append a Column for Updated At 
```
public function appendColumns(): array
{
    return [
        Column::make('Updated At', 'updated_at'),
    ];
}
```

**Comment by lrljoe** [comment]
Noting that v3 is now released as stable 3.1.0

**Comment by jordivela007** [comment]
resources / views / vendor / livewire-tables / components / tools / filters / text-field.blade.php

Undefined variable $component

i upgraded from livewire 2 to livewire 3. What's wrong?

**Reply by slossetti** [reply]
Same issue here.

**Reply by lrljoe** [reply]
If you previously published the views, remove them.

**Reply by slossetti** [reply]
Ohhh thanks! I forgot that. I delete it. Now is working!
